src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -92,16 +111,107 @@ protected override Payload ConstructHeadersPayload(BlockLocator locator, uint256
-                this.chain, this.initialBlockDownloadState, this.consensusManager, this.peerBanning, this.loggerFactory, this.network);
+                this.chain,
+                this.initialBlockDownloadState,
+                this.consensusManager,
+                this.peerBanning,
+                this.loggerFactory,
+                this.network,
+                this.chainState);
+        }
+
+        /// <summary>
+        /// Determines whether the specified peer supports Proven Headers and PH has been activated.
+        /// </summary>
+        /// <param name="peer">The peer.</param>
+        /// <returns>
+        ///   <c>true</c> if is peer is PH enabled; otherwise, <c>false</c>.
+        /// </returns>
+        private bool IsPeerPhEnabled(INetworkPeer peer)
+        {
+            return peer.Version >= NBitcoin.Protocol.ProtocolVersion.PROVEN_HEADER_VERSION;
+        }
+
+        /// <summary>
+        /// Determines whether the specified peer is Whitelisted.
+        /// </summary>
+        /// <param name="peer">The peer.</param>
+        /// <returns>
+        ///   <c>true</c> if the specified peer is Whitelisted; otherwise, <c>false</c>.
+        /// </returns>
+        private bool IsPeerWhitelisted(INetworkPeer peer)
+        {
+            return peer.Behavior<IConnectionManagerBehavior>()?.Whitelisted == true;
+        }
+
+
+        private bool IsProvenHeaderActivated()
+        {
+            if (this.network.Consensus.Options is PosConsensusOptions options)
+            {
+                long currentHeight = this.chainState.ConsensusTip.Height;
+                return (options.ProvenHeadersActivationHeight > 0) && (currentHeight >= options.ProvenHeadersActivationHeight);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Builds the GetHeadersPayload.
+        /// </summary>
+        /// <returns>The GetHeadersPayload instance.
+        /// If the peer can serve PH, GetProvenHeadersPayload is returned, otherwise if it's a legacy peer but it's whitelisted,
+        /// GetHeadersPayload is returned.
+        /// If the attached peer is a legacy peer and it's not whitelisted, returns null.
+        /// </returns>
+        protected override GetHeadersPayload BuildGetHeadersPayload()
+        {
+            INetworkPeer peer = this.AttachedPeer;
+
+            if (this.IsProvenHeaderActivated())
+            {
+                if (this.IsPeerPhEnabled(peer))
+                {
+                    return new GetProvenHeadersPayload()
+                    {
+                        BlockLocator = (this.ExpectedPeerTip ?? this.consensusManager.Tip).GetLocator(),
+                        HashStop = null
+                    };
+                }
+                // If the peer doesn't supports PH but it's whitelisted, issue a standard GetHeadersPayload
+                else if (IsPeerWhitelisted(peer) || this.IsProvenHeaderActivated())
```

Comment:
`this.IsProvenHeaderActivated()` will always be true here.

Reply:
yep!
```suggestion
                else if (IsPeerWhitelisted(peer))
```